### PR TITLE
Fix problems in .start() algorithm.

### PR DESCRIPTION
**Pull request by jan-ivar**
Fixes https://github.com/w3c/mediacapture-record/issues/55 and part of https://github.com/w3c/mediacapture-record/issues/54.

### CODE REVIEW

File: MediaRecorder.html [hunk 1]
```diff
@@ -231,7 +239,7 @@
...
+   <p><dfn data-dfn-type="dfn" id="fire-error-event">Firing an error event named <var>e</var></dfn> with an <code>Error</code> <var>error</var> means that an event with the name <var>e</var>, which does not bubble (except where otherwise stated) and is not cancelable (except where otherwise stated), and which uses the <code><a>ErrorEvent</a></code> interface with the <code><a href="#widl-ErrorEvent-error">error</a></code> attribute set to <var>error</var>, MUST be created and dispatched at the given target. If no <code>Error</code> object is specified, the <code><a href="#widl-ErrorEvent-error">error</a></code> attribute defaults to null.</p>
```
Comment (yell0wd0g) [inline MediaRecorder.html]: s/MUST/<em title="must" class="rfc2119">must</em>/

File: MediaRecorder.html [hunk 2]
```diff
@@ -92,22 +92,30 @@
...
+           <li>If at any point the stream's isolation properties change so that MediaRecorder is no longer allowed access to it, the UA <em title="must" class="rfc2119">must</em> immediately stop gathering data, discard any data that it has gathered, and queue a task, using the DOM manipulation task source, that <a href="#track-ended">fires an error event</a> with the name <code>SecurityError</code>, immediately followed by a <code>stop</code> event.
+           <li>If the UA at any point is unable to continue gathering data for reasons other than isolation properties, it <em title="must" class="rfc2119">must</em> stop gathering data, let <var>blob</var> be the Blob of collected data, and queue a task, using the DOM manipulation task source, that <a href="#track-ended">fires an error event</a> with the name <code>UnknownError</code>, followed by a <code>dataavailable</code> event containing <var>blob</var>, follwed by a <code>stop</code> event.</li>
```
Comment (yell0wd0g) [inline MediaRecorder.html]: s/follwed/followed/

File: MediaRecorder.html [hunk 3]
```diff
@@ -92,22 +92,30 @@
...
-      <li>Then fire a <code>dataavailable</code> event containing the Blob of data.</li>
-      <li>Finally, fire a <code>stop</code> event.</li>
-     </ol>
+    <dd>When a <code>MediaRecorder</code> object’s <code>start()</code> method is invoked, the UA <em title="must" class="rfc2119">must</em> run the following steps:
+      <ol>
+        <li>Let <var>timeslice</var> be the method's first argument, if provided, or <code>undefined</code>.</li>
+        <li>If <code>state</code> is not "inactive", throw an <code>InvalidStateError</code> DOMException and abort these steps.</li>
+        <li>If the stream's isolation properties disallow access from this MediaRecorder, throw a <code>SecurityError</code> DOMException and abort these steps.</li>
+        <li>Set <code>state</code> to "recording", and run the following steps in parallel:</li>
```
Comment (yell0wd0g) [inline MediaRecorder.html:100]: Do you need to specify "in parallel" ?

File: MediaRecorder.html [hunk 4]
```diff
@@ -92,22 +92,30 @@
...
+           <li>If <var>timeslice</var> is not <code>undefined</code>, then once a minimum of <var>timeslice</var> milliseconds of data have been collected, or some minimum time slice imposed by the UA, whichever is greater, let <var>blob</var> be the Blob of collected data, start gathering a new Blob of data, and queue a task, using the DOM manipulation task source, that fires a <code>dataavailable</code> event containing <var>blob</var>.</li>
+           <li>If all recorded tracks become <a href="/mediacapture-main/getusermedia.html#track-ended">ended</a>, then stop gathering data, let <var>blob</var> be the Blob of collected data, and queue a task, using the DOM manipulation task source, that runs the following steps:</li>
```
Comment (yell0wd0g) [inline MediaRecorder.html:106]: I'd like @foolip to take a look at this changes and see if the line up with other media specs.

File: MediaRecorder.html (hunk 3 above)
Comment (jan-ivar) [inline MediaRecorder.html:100]: @miguelao Fyi it is defined [here](https://html.spec.whatwg.org/#in-parallel). How do you want to handle it? FWIW the webrtc spec uses "in parallel" without defining it.

File: MediaRecorder.html (hunk 4 above)
Comment (jan-ivar) [inline MediaRecorder.html:106]: This particular line of change was meant to be neutral (same behavior as before), but let me know if I messed something up.

File: MediaRecorder.html [hunk 5]
```diff
@@ -92,22 +92,30 @@
...
+           <li>If the UA at any point is unable to continue gathering data for reasons other than isolation properties, it <em title="must" class="rfc2119">must</em> stop gathering data, let <var>blob</var> be the Blob of collected data, and queue a task, using the DOM manipulation task source, that <a href="#track-ended">fires an error event</a> with the name <code>UnknownError</code>, followed by a <code>dataavailable</code> event containing <var>blob</var>, follwed by a <code>stop</code> event.</li>
+           <li>If <var>timeslice</var> is not <code>undefined</code>, then once a minimum of <var>timeslice</var> milliseconds of data have been collected, or some minimum time slice imposed by the UA, whichever is greater, let <var>blob</var> be the Blob of collected data, start gathering a new Blob of data, and queue a task, using the DOM manipulation task source, that fires a <code>dataavailable</code> event containing <var>blob</var>.</li>
```
Comment (foolip) [inline MediaRecorder.html:105]: If we're being picky, the "fired" here needs to be linked to something in the style of https://html.spec.whatwg.org/multipage/webappapis.html#fire-a-synthetic-mouse-event that says what the event interface is and what the various flags are set to.

File: MediaRecorder.html (hunk 4 above)
Comment (foolip) [inline MediaRecorder.html:106]: This looks good to me, using the DOM manipulation task source is probably the right choice as long as it's never possible that the events which have already been queued should be removed from the queue. But... I'm not certain that's the case. If one calls stop() and start() in quick succession with unfortunate timing, one could get the dataavailable and stop events from before the stop and have no way of knowing what they mean.

The way that media elements deal with a similar situation is to have a task source per instance, the "media element task source". It's kind of annoying, but if you have that kind of lifecycle then I don't see another way to make the API robust.

File: MediaRecorder.html (hunk 3 above)
Comment (yell0wd0g) [inline MediaRecorder.html:100]: Nay, it's ok like that.

File: MediaRecorder.html (hunk 4 above)
Comment (jan-ivar) [inline MediaRecorder.html:106]: @foolip thanks for looking at this, as this exceeds my understanding of task sources. My experience is mostly with MediaCapture and Streams and WebRTC, which both use the phrase "queue a task" liberally and without specifying a source. I only specified the _DOM manipulation task source_ here since it was previously mentioned. Happy to put in whatever we think is best.

File: MediaRecorder.html (hunk 5 above)
Comment (jan-ivar) [inline MediaRecorder.html:105]: Since the "fire event" language seems orthogonal to this PR, and its use in this spec extends way beyond the start algorithm, could this be dealt with as a separate issue?

Comment (yell0wd0g) [inline MediaRecorder.html:105]: SG, please open a parallel issue so we won't forget.

**Comment by yell0wd0g** [comment]
@jan-ivar if you can correct those minor ortographical/upper-lowercase nits I think this PR is good to be merged.

File: MediaRecorder.html (hunk 4 above)
Comment (foolip) [inline MediaRecorder.html:106]: I think it either has to be the DOM manipulation task source or a task source that's unique per MediaRecorder object. That depends on how you want the stop() and start() situation to work, any thoughts? I can file a new issue if you like.

Comment (jan-ivar) [inline MediaRecorder.html:106]: @foolip new issue would be great thanks. I don't think I have a strong preference, though I don't fully see how picking a task source per recorder changes the fundamental problem you mention that one can call `stop` and `start` multiple times in a row synchronously, and the ensuing chaos of subsequent events this triggers.

**Comment by yell0wd0g** [comment]
@jan-ivar this change LGTM, could you somehow make it against `gh-pages` branch? (so it's automatically published).

**Comment by jan-ivar** [comment]
@miguelao I can, but when I tried that, I noticed `master` and `gh-pages` both appear to have changes in them since they split.

`gh-pages`:

```
8dc6b12 Remove anti-slash from after_script in .travis.yml Otherwise, Travis CI reports too many arguments for test
2231040 Use the right syntax to use check values of the TRAVIS_BRANCH and TRAVIS_PULL_REQUEST variables befaore calling echidna (thx to http://blog.tgrrtt.com/exploring-the-travisci-configuration-file#travis-amp-bash-scripts_2)
a85a28d Removing [[  == 'false' ]] as teh full line was not a valid YAML syntax see: https://github.com/travis-ci/travis-ci/issues/5933
c174060 Markup validation fix
08da5da Editorial comment to test push-to-WD
362bfbf Add Travis support to gh-pages branch
782a6bc Remove comment from 1st line of manifest
ad54121 removed inexistent <a></a>
b30713c Added dated version v20160203
c3ffeaf Remove MediaRecorder Properties Section, which was a TODO. More comments in {audio,video}BitsPerSecond reader
03f2fde Merge pull request #49 from miguelao/master
```

`master`:

```
df3c2c8 Fix existing typos in text that was touched.
2cb9f62 Fix problems in .start() algorithm.
45a31a9 Merge pull request #58 from dontcallmedom/webidl-typo
7c0b571 Fix WebIDL typo
6088a93 Do not trigget automatic publishing if was a new pull request (we only want to launch publication when the PR are merged)
edff2e9 Merge pull request #51 from w3c/echidna
88e5627 remove extra space character
6b42d30 Fix DECISION link that relates to the group agreeing to set up automatic publishing
2e949f8 Trigger echidna automatic publications from Travis
6d0a4e9 Use HTTPS for w3c.github.io and some w3.org links
e460bd1 Fix validation issue.
80581cc Merge pull request #50 from miguelao/master
c3ffeaf Remove MediaRecorder Properties Section, which was a TODO. More comments in {audio,video}BitsPerSecond reader
03f2fde Merge pull request #49 from miguelao/master
```

Should this perhaps be resolved first?

**Comment by yell0wd0g** [comment]
Hmm probably we should stick to one single model, either we work all the time in `gh-pages`, or we use `master` as temporary branch that dumps a bundle of commits onto `gh-pages`. I'd just do the former --  note that the 'Editor's Draft' version points to `gh-pages` and now, with Echidna, so does the 'last published version'.

@dontcallmedom, @vivienlacourba, the commits @jan-ivar mentions are mostly yours and you understand better the publications process, what would you recomment?

**Comment by vivienlacourba** [comment]
Hi @miguelao,

Your proposal to only use `gh-pages` sounds like the easiest and safest one.
Steps to move forward:
1) I'll integrate in the `gh-pages` branch the changes that are today only `master`
2) Mark `gh-pages` the default branch for this repo
3) Then ask @jan-ivar to rebase this PR against `gh-pages`
4) delete the `master` branch

**Comment by vivienlacourba** [comment]
@jan-ivar the above 1) and 2) have been done. can you make you PR against the `gh-pages` branch ?

**Comment by jan-ivar** [comment]
Done, thanks!

**Comment by vivienlacourba** [comment]
Thx @jan-ivar !
@miguelao I let you take care of deleting the `master` branch

File: MediaRecorder.html (hunk 5 above)
Comment (foolip) [inline MediaRecorder.html:105]: Finally: https://github.com/w3c/mediacapture-record/issues/69
